### PR TITLE
switch the ADAL android to Android Studio 

### PR DESCRIPTION
**Pull request by heidijinxujia**
• configure Findbugs, PMD, Android Lint and Checkstyle into the gradle build
• create the jar and aar files
• unify the names of the output
• update the ADAL to 1.1.16
• remove Travis

### CODE REVIEW

File: testapp/build.gradle [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-    buildToolsVersion "23.0.2"
+    compileSdkVersion 21
+    buildToolsVersion "21.1.2"
     useLibrary 'org.apache.http.legacy'
 
     defaultConfig {
         applicationId "com.microsoft.aad.testapp"
         minSdkVersion 18
-        targetSdkVersion 23
+        targetSdkVersion 21
```
Comment (imakhalova) [inline testapp/build.gradle]: decreasing sdk version?

File: sample/src/main/java/com/microsoft/aad/adal/sample/Constants.java [hunk 2]
```diff
@@ -32,19 +32,19 @@
     
     // AAD PARAMETERS
     // https://login.windows.net/tenantInfo
-    static final String AUTHORITY_URL = "https://login.windows.net/common";
+    static final String AUTHORITY_URL = "https://your.authority.url.here";
```
Comment (imakhalova) [inline sample/src/main/java/com/microsoft/aad/adal/sample/Constants.java]: revert this file, it's actually nice to have working app with valid values

File: sample/sample.iml [hunk 3]
```diff
@@ -65,34 +65,25 @@
...
-    <orderEntry type="library" exported="" name="appcompat-v7-23.1.0" level="project" />
+    <orderEntry type="library" exported="" name="appcompat-v7-21.0.3" level="project" />
     <orderEntry type="library" exported="" name="gson-2.2.4" level="project" />
-    <orderEntry type="library" exported="" name="support-v4-23.1.0" level="project" />
+    <orderEntry type="library" exported="" name="support-annotations-21.0.3" level="project" />
+    <orderEntry type="library" exported="" name="support-v4-21.0.3" level="project" />
+    <orderEntry type="library" exported="" name="recyclerview-v7-21.0.3" level="project" />
     <orderEntry type="module" module-name="adal" exported="" />
-    <orderEntry type="library" exported="" name="org.apache.http.legacy-android-23" level="project" />
+    <orderEntry type="library" exported="" name="C__Users_jinjia_AppData_Local_Android_sdk_platforms_android_21_optional_org_apache_http_legacy_jar-android-21" level="project" />
```
Comment (imakhalova) [inline sample/sample.iml]: ? pretty sure that no one has this lib)

File: sample/sample.iml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (imakhalova) [inline sample/sample.iml]: I think this iml  files don't need to be checked in at all, they are generated

File: adal/src/androidTest/java/com/microsoft/aad/test/AndroidTestHelper.java [hunk 5]
```diff
@@ -56,7 +56,7 @@ protected void setUp() throws Exception {
 
         // ADAL is set to this signature for now
         PackageInfo info = getInstrumentation().getContext().getPackageManager()
-                .getPackageInfo("com.microsoft.aad.testapp", PackageManager.GET_SIGNATURES);
+                .getPackageInfo("com.microsoft.aad.adal.testapp", PackageManager.GET_SIGNATURES);
```
Comment (imakhalova) [inline adal/src/androidTest/java/com/microsoft/aad/test/AndroidTestHelper.java]: put packageName in some static final

File: adal/src/androidTest/java/com/microsoft/aad/test/AssertUtils.java [hunk 6]
```diff
@@ -16,7 +16,7 @@
 // See the Apache License, Version 2.0 for the specific language
 // governing permissions and limitations under the License.
 
-package com.microsoft.aad.adal;
+package com.microsoft.aad.test;
```
Comment (imakhalova) [inline adal/src/androidTest/java/com/microsoft/aad/test/AssertUtils.java]: tests should stay at the same package with project

File: adal/build.gradle [hunk 7]
```diff
@@ -65,7 +66,8 @@ dependencies {
     androidTestCompile 'org.mockito:mockito-core:1.10.19'
     androidTestCompile 'com.google.dexmaker:dexmaker:1.2'
     androidTestCompile 'com.google.dexmaker:dexmaker-mockito:1.2'
-    compile 'com.android.support:appcompat-v7:21.0.2'
+    compile 'com.android.support:appcompat-v7:21.0.3'
+    compile 'com.android.support:recyclerview-v7:21.0.3'
```
Comment (imakhalova) [inline adal/build.gradle]: why do you need recyclerView? there is no usage in lib

File: adal/adal.iml [hunk 8]
```diff
@@ -65,15 +65,14 @@
...
-      <excludeFolder url="file://$MODULE_DIR$/build/docs" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/annotations" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/assets" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/blame" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/bundles" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/classes" />
       <excludeFolder url="file://$MODULE_DIR$/build/intermediates/dependency-cache" />
-      <excludeFolder url="file://$MODULE_DIR$/build/intermediates/exploded-aar/com.android.support/appcompat-v7/23.1.0/jars" />
-      <excludeFolder url="file://$MODULE_DIR$/build/intermediates/exploded-aar/com.android.support/support-v4/23.1.0/jars" />
+      <excludeFolder url="file://$MODULE_DIR$/build/intermediates/exploded-aar/com.android.support/appcompat-v7/21.0.2/jars" />
```
Comment (imakhalova) [inline adal/adal.iml]: leave it as 23 and upgrade everything to 23

File: testapp/build.gradle (hunk 1 above)
Comment (heidijinxujia) [inline testapp/build.gradle]: Hi @imakhalova, I am trying to make this sdk version same with the current adal master branch. Do you prefer it to be updated to v23？

File: sample/sample.iml (hunk 3 above)
Comment (weijjia) [inline sample/sample.iml]: +1 also we should have user specific thing here.

File: testapp/build.gradle (hunk 1 above)
Comment (imakhalova) [inline testapp/build.gradle]: Sure, kill 2 birds with 1 stone, its actually very good practice always keep the latest stable android sdk in your project.

Comment (weijjia) [inline testapp/build.gradle]: we should keep all the sample apps target as 21 unless we're ready to update to 23.

File: adal/adal.iml (hunk 8 above)
Comment (heidijinxujia) [inline adal/adal.iml]: OK I am working on this. Thank you for your reviews, @imakhalova !

File: testapp/build.gradle (hunk 1 above)
Comment (imakhalova) [inline testapp/build.gradle]: You're always ready for 23. It's like saying are you clients ready for new version of adal where you were fixing bugs and adding features?

File: sample/src/main/java/com/microsoft/aad/adal/sample/Constants.java (hunk 2 above)
Comment (heidijinxujia) [inline sample/src/main/java/com/microsoft/aad/adal/sample/Constants.java]: OK

File: testapp/build.gradle (hunk 1 above)
Comment (imakhalova) [inline testapp/build.gradle]: May you reveal please what is the purpose of keeping old sdk?

**Comment by imakhalova** [comment]
Did you take a look how to switch travis to use gradle to build project?

File: sample/build.gradle [hunk 9]
```diff
@@ -28,5 +28,6 @@ dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
     testCompile 'junit:junit:4.12'
     compile 'com.android.support:appcompat-v7:21.0.3'
-    compile 'com.microsoft.aad:adal:1.1.16'
-}
+    //compile 'com.microsoft.aad:adal:1.1.16'
+    compile project(':adal')
```
Comment (weijjia) [inline sample/build.gradle]: why switching back to local lib?

File: sample/build.gradle [hunk 10]
```diff
@@ -3,7 +3,7 @@ apply plugin: 'com.android.application'
 android {
     compileSdkVersion 21
     buildToolsVersion "21.1.1"
-    //useLibrary 'org.apache.http.legacy'
+    //useLibrary 'org.apache.http.legacy' //FOR sdk 23
```
Comment (weijjia) [inline sample/build.gradle]: if you don't need them, can you please delete it?

**Comment by weijjia** [comment]
For switching travis to gradle, i think we decide to skip it for now? @heidijinxujia correct me if i'm wrong.

**Comment by weijjia** [comment]
@heidijinxujia Please make the copy right header is updated to MIT license. It's already updated in dev, if you do a

**Comment by heidijinxujia** [comment]
@weijjia  Yes, I closed the travis checks for this branch.

**Comment by heidijinxujia** [comment]
@weijjia  Hi, I am afraid you didn't finish the last comment, do you? :)

File: adal/build.gradle [hunk 11]
```diff
@@ -1,3 +1,13 @@
+apply plugin: 'com.android.library'
+// This plugin publishes adal in to the local maven repo
+apply plugin: 'com.github.dcendents.android-maven'
+apply plugin: 'findbugs'
+apply plugin: 'pmd'
+apply plugin: 'checkstyle'
+
+group = 'com.microsoft.aad'
+
+
```
Comment (weijjia) [inline adal/build.gradle]: nit: extra line

**Comment by weijjia** [comment]
Yup, sorry about it. Already updated. BTW, can we separate the file merging and gradle config? Can we have the PR only contain the gradle build? After we finish the PR, you can go ahead merging the latest changes form dev to this branch. What do you think?

**Comment by heidijinxujia** [comment]
OK. So separate the gradle congif with the code update into different PRs, right? @weijjia

**Comment by weijjia** [comment]
You only need to send the PR for gradle config, and then go ahead merging the code.

**Comment by heidijinxujia** [comment]
Code cannot be merged automatically from dev, because they are in totally different file structures (dev is maven build, whereas this is gradle build) with different config files.

**Comment by heidijinxujia** [comment]
Change the branch structure.

**Pull request closed without merging**